components/pluralrules/src/rules/resolver.rs

```diff
@@ -0,0 +1,86 @@
+use super::ast;
+use crate::operands::PluralOperands;
+
+/// Function used to test [`Condition`] against [`PluralOperands`] to identify
+/// the appropriate [`PluralCategory`].
+///
+/// # Examples
+///
+/// ```
+/// use icu_pluralrules::PluralOperands;
+/// use icu_pluralrules::rules::{parse, test_condition};
+///
+/// let operands = PluralOperands::from(5_usize);
+/// let condition = parse(b"i = 4..6")
+///     .expect("Failde to parse a rule.");
+///
+/// assert_eq!(test_condition(&condition, &operands), true);
+/// ```
+///
+/// [`PluralCategory`]: ../enum.PluralCategory.html
+/// [`PluralOperands`]: ../struct.PluralOperands.html
+/// [`Condition`]: ../rules/ast/struct.Condition.html
+pub fn test_condition(condition: &ast::Condition, operands: &PluralOperands) -> bool {
+    condition.0.is_empty() || condition.0.iter().any(|c| test_and_condition(c, operands))
+}
+
+fn test_and_condition(condition: &ast::AndCondition, operands: &PluralOperands) -> bool {
+    condition.0.iter().all(|r| test_relation(r, operands))
+}
+
+fn test_relation(relation: &ast::Relation, operands: &PluralOperands) -> bool {
+    calculate_expression(&relation.expression, operands)
+        .map(|exp| test_range(&relation.range_list, exp, &relation.operator))
+        .unwrap_or(false)
+}
+
+// At runtime CLDR only works with integer values.
+// If a rule is asking for `n` it will compare it to
+// a list of integers, and expect it to not match if
+// there is any fractional part.
+//
+// That allows us to play a trick - we already have an
+// integer portion of the number as operand `i`.
+//
+// This function returns `None` to indicate that the value should
+// not match anything which will happen if expression asks for operand `n`
+// which has a fractional part, or if it uses a modulus `0`.
+fn calculate_expression(expression: &ast::Expression, operands: &PluralOperands) -> Option<u64> {
+    let value = match expression.operand {
+        ast::Operand::N => {
+            if operands.n.fract() < f64::EPSILON {
+                operands.i
+            } else {
+                return None;
```

Comment:
Optional: `debug_assert!` here so that if your precondition ever fails to be true, it will fail loudly.

Reply:
What precondition? Both branches are viable and used and tested.

Reply:
OK.  The trick is that in [UTS 35 Part 2 Section 5.1](https://unicode.org/reports/tr35/tr35-numbers.html#Plural_rules_syntax), we have:

```
relation      = is_relation | in_relation | within_relation 
is_relation   = expr 'is' ('not')? value
in_relation   = expr (('not')? 'in' | '=' | '!=') range_list
within_relation = expr ('not')? 'within' range_list
expr          = operand (('mod' | '%') value)?
operand       = 'n' | 'i' | 'f' | 't' | 'v' | 'w'
range_list    = (range | value) (',' range_list)*
range         = value'..'value
value         = digit+
```

So, all values, even `n`, are defined as whole integers via `value = digit+`.

Consider rephrasing your comment to point this out.  The way I read it, "at runtime CLDR only works with integer values", makes it sound like you are relying on some invariant about locale data, when you are actually encoding an assumption from the AST.